source/building-your-own-javascript-components/index.html.md.erb

```diff
@@ -0,0 +1,283 @@
+
+```js
+class MyComponent extend Component{
+  static property = “value”
+}
+```
+
+Otherwise, you can add these properties to your component’s class after its definition:
+
+```js
+class MyComponent extends Component {}
+MyComponent.property = “value”
+```
+
+### Defining the `moduleName` property
+
+Components extending the `Component` class need to define a `moduleName` [static property](#defining-static-properties-on-components):
+
+```js
+class MyComponent extend Component{
+  static moduleName = “app-my-component”
+}
+```
+
+Doing this helps:
+- [`createAll` initialise the component](#using-createall-with-your-components)
+- support consistent error messaging
+
+### Customising initialisation of components
+
+When your component is initialised, it'll need to perform some tasks, such as adding event listeners. You can add these tasks to your component’s `constructor` after calling `super` to set up the base `Component`’s default behaviour.
+
+```js
+class MyComponent extends Component {
+  static moduleName = ‘app-my-component’
+
+  constructor($root) {
+    super($root)
+
+    // Run component specific initialisation here for example
+    this.$root.addEventListener(‘click’, this.handleClick.bind(this))
+  }
+
+  handleClick(event) {
+    // Handle click inside the component’s root
+  }
+}
+```
+
+### Accessing the root element in the component
+
+Within the component’s methods and its constructor, you can access the root element the component was initialised on using `this.$root`.
+
+```js
+class MyComponent extends Component {
+  static moduleName = ‘app-my-component’
+
+  constructor($root) {
+    super($root)
+
+    // Run component specific initialisation here, for example:
+    this.$root.addEventListener(‘click’, this.handleClick.bind(this))
+  }
+
+  handleClick(event) {
+    this.$root.classList.add(‘app-my-component--clicked’)
+  }
+}
+```
+
+#### Customising the expected type of the root element
+
+The `Component` class will verify that the root element is of the right type (by default, `HTMLElement`). You can set an `elementType` [static property](#defining-static-properties-on-components) on your component to define a specific type to check against:
+
+```js
+/**
+ * @extends Component<HTMLAnchorElement> 
+ */
+class LinkEnhancement extends Component {
+  static moduleName = ‘app-link-enhancement’
+
+  // This will make sure that the component can only be initialised
+  // on HTMLAnchorElement
+  static rootElementType = HTMLAnchorElement
+
+  constructor($root) {
+    super($root);
+
+    // this.$root is an `HTMLAnchorElement` and you can access specific properties
+    // like `href` or `hash`
+  }
+}
+```
+
+#### Using `this.$root` with TypeScript
+
+As `govuk-frontend` does not provide type definitions in its package, TypeScript might show the following error message: 
+
+<q>“Property ‘$root’ does not exist on type &lt;NAME_OF_YOUR_CLASS&gt;”.</q>
```

Comment:
The <q> tag already provides quotation marks, so the quotation marks have doubled up and looks like this right now:

> "“Property ‘$root’ does not exist on type <NAME_OF_YOUR_CLASS>”."

Can we remove the quotation marks on this code line? If that's not appropriate, and we do want nested quotation marks, converting them to single quotes (') would be the correct punctuation.

Reply:
My bad for not picking that up when I brought in your suggestion, I'll remove the extra quotes 😊

Reply:
```suggestion
<q>Property ‘$root’ does not exist on type &lt;NAME_OF_YOUR_CLASS&gt;.</q>
```